### PR TITLE
Fix metrics names collision

### LINKED ISSUE #28: [BUG] - Unable to do model selection on a decreasing metric
**Describe the bug**
I expect to be able to do model selecting, using the classifier loss as metric. Actually, I find only a JSON with this content:

> cat ~/log/on/best/loss/loss.json 

```
{
    "loss": "-inf",
    "step": "0"
}
```

**Expected behavior**

I should find in the logdir, the best folder containing the JSON + the checkpoint files of the best model, wrt the chosen metric.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```

def get_model():
    """Create a new autoencoder tf.keras.Model."""
    autoencoder = Autoencoder(
        (64, 64),
        (4, 4),
        kernel_size=3,
        initial_filters=16,
        filters_cap=64,
        encoding_dimension=50,
        channels=3,
    )

    # encoding, representation = autoencoder(input)
    inputs = tf.keras.layers.Input(shape=(64, 64, 3))
    _, reconstruction = autoencoder(inputs)
    model = tf.keras.Model(inputs=inputs, outputs=reconstruction)
    return model


def _train(dataset: tf.data.Dataset, logdir: Path):
    reconstruction_error = ClassifierLoss(tf.keras.losses.MeanSquaredError())
    autoencoder = get_model()

    ClassifierTrainer(
        model=autoencoder,
        optimizer=tf.optimizers.Adam(1e-4),
        loss=reconstruction_error,
        metrics=[ashpy.metrics.ClassifierLoss(model_selection_operator=operator.lt)],
        logdir=str(logdir),
        epochs=100,
    )(dataset, dataset)

```

In short, the initialization of the JSON is wrong when I want to select a model whose metric is decreasing.

### PR DESCRIPTION
**Pull request by mr-ubik**
## Description

The bug was due to the collision of the names of the metrics. It has been
corrected by adding a collision detector to the
Trainer._validate_metrics(). Now if two metrics have the same name
a ValueError exception will be thrown during trainer instantiation.

Fixes #28 

## Type of change

Please delete options that are not relevant.

- [x] Bugfix (i.e., a non-breaking change which fixes an issue)
- [x] Breaking change (i.e., a fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

- Enhanced `test_metrics.py`
- `tox` test suite

### Dependencies

List any dependencies that are required for this change.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by EmanueleGhelfi** [comment]
@mr-ubik please fix the issues with CodeFactor

**Review by galeone — APPROVED** [review]

**Review by galeone — APPROVED** [review]
